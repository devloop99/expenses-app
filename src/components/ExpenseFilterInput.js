import React from "react";
import { connect } from "react-redux";
import { setText, SortByAmount, SortByDate } from "../actions/filter";
import { DateRangePicker } from "react-dates";
import moment from "moment";
import { setEndDate, setStartDate } from "../actions/filter";

class ExpenseFilterInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      option: "date",
      isFocused: undefined,
      startDate: moment().startOf("month"),
      endDate: moment().endOf("month"),
    };
  }
  changeHandler = (e) => {
    const option = e.target.value;
    this.setState(() => ({ option }));
    if (option === "date") {
      this.props.dispatch(SortByDate());
    } else if (option === "amount") {
      this.props.dispatch(SortByAmount());
    }
  };
  dateChange = (focusedInput) => {
    this.setState({ isFocused: focusedInput });
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={(e) => {
            this.props.dispatch(setText(e.target.value));
          }}
        />
        <select value={this.state.option} onChange={this.changeHandler}>
          <option value="amount">Amount</option>
          <option value="date">Date</option>
        </select>
        <DateRangePicker
          startDate={this.state.startDate} // momentPropTypes.momentObj or null,
          endDate={this.state.endDate} // momentPropTypes.momentObj or null,
          endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
          onDatesChange={({ startDate, endDate }) => {
            this.setState(() => ({ startDate }));
            this.setState(() => ({ endDate }));
            this.props.dispatch(
              setStartDate(startDate ? startDate.valueOf() : 0)
            );
            this.props.dispatch(
              setEndDate(endDate ? endDate.valueOf() : undefined)
            );
          }} // PropTypes.func.isRequired,
          focusedInput={this.state.isFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
          onFocusChange={this.dateChange} // PropTypes.func.isRequired,
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

const ExpenseFilterInputConnected = connect(mapStateToProps)(
  ExpenseFilterInput
);

export default ExpenseFilterInputConnected;
