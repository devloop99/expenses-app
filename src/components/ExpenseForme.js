import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

export default class ExpenseForme extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: !!this.props.expense ? this.props.expense.description : "",
      amount: !!this.props.expense ? this.props.expense.amount : "",
      note: !!this.props.expense ? this.props.expense.note : "",
      createdAt: !!this.props.expense
        ? moment(this.props.expense.createdAt)
        : moment(),
      isFocused: false,
    };
  }
  focusChange = ({ focused }) => {
    this.setState(() => ({ isFocused: focused }));
  };
  descriptionChange = (e) => {
    let description = e.target.value;

    this.setState(() => ({ description }));
  };
  noteChange = (e) => {
    const note = e.target.value;

    this.setState(() => ({ note }));
  };
  amountChange = (e) => {
    let amount = e.target.value;
    if (amount.match(/^\d*(\.\d{0,2})?$/)) {
      amount = e.target.value;
      this.setState(() => ({ amount }));
    }
  };
  onSubmit = () => {
    if (this.state.description)
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount)
          ? parseFloat(this.state.amount)
          : 0,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
      });
  };
  render() {
    return (
      <from>
        <input
          type="text"
          placeholder="Description"
          value={this.state.description}
          onChange={this.descriptionChange}
          autoFocus
        />
        <input
          type="text"
          onChange={this.amountChange}
          value={this.state.amount}
          placeholder="Amount"
        />
        <textarea
          placeholder="your note"
          value={this.state.note}
          onChange={this.noteChange}
        ></textarea>
        <SingleDatePicker
          date={this.state.createdAt} // momentPropTypes.momentObj or null
          onDateChange={(createdAt) => this.setState({ createdAt })} // PropTypes.func.isRequired
          focused={this.state.isFocused} // PropTypes.bool
          numberOfMonths={1}
          isOutsideRange={() => false}
          onFocusChange={this.focusChange} // PropTypes.func.isRequired
        />
        <button onClick={this.onSubmit}>Add expense</button>
      </from>
    );
  }
}
