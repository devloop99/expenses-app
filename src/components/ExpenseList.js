import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./expenseListItem";
import getVisible from "../selectors/expenses";

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length > 0 ? (
      props.expenses.map((el) => <ExpenseListItem key={el.id} {...el} />)
    ) : (
      <p>No expense yet ...</p>
    )}
  </div>
);
const mapStateToProps = (state) => ({
  expenses: getVisible(state.expenses, state.filters),
});

export default connect(mapStateToProps)(ExpenseList);
