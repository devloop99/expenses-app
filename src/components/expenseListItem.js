import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";

//5000
//500.55
//formatting function
export const formatNum = (num) => {
  let arr = num.toFixed(2).split(".");
  if (arr[0].length > 3) {
    let den = arr[0];
    arr[0] =
      den.slice(0, den.length - 3) +
      "," +
      den.slice(den.length - 3, den.length);
  }

  return "$" + arr[0] + "." + arr[1];
};

export const ExpenseListItem = ({ description, amount, createdAt, id }) => {
  return (
    <dive>
      <Link to={"/edit/" + id}>
        <h3>{description}</h3>
      </Link>

      <p>
        Amount : {formatNum(amount)}
        <br />
        Date : {moment(createdAt).format("MMM, Do YYYY")}
      </p>
    </dive>
  );
};

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};

const ExpenseListItemConnected = connect(mapStateToProps)(ExpenseListItem);

export default ExpenseListItemConnected;
