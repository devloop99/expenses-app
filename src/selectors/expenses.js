const getVisible = (expense, { text, sortBy, startDate, endDate }) => {
  const visible = expense
    .filter((el) => {
      const startMatch =
        typeof startDate !== "number" || el.createdAt >= startDate;

      const endMatch = typeof endDate !== "number" || el.createdAt <= endDate;
      const textMatch =
        el.description.toLowerCase().includes(text) || el.note.includes(text);

      return startMatch && endMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return -(a.createdAt - b.createdAt);
      } else if (sortBy === "amount") {
        return b.amount - a.amount;
      }
    });

  return visible;
};

export default getVisible;
