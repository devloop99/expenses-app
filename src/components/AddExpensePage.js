import React from "react";
import ExpenseForme from "./ExpenseForme";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expense";

class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.dispatch(startAddExpense(expense));
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h2>CREATE EXPENSE :</h2>
        <ExpenseForme onSubmit={this.onSubmit} />
      </div>
    );
  }
}

const AddExpensePageConnected = connect()(AddExpensePage);

export default AddExpensePageConnected;
