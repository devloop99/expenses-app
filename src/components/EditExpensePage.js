import React from "react";
import ExpenseForme from "./ExpenseForme";
import { connect } from "react-redux";
import {
  startEditExpense as editExpense,
  startRemoveExpense as removeExpense,
} from "../actions/expense";

class EditExpensePage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      expense: props.expenses.find((el) => el.id === props.match.params.id),
    };
  }
  onSubmit = (expense) => {
    const id = this.props.match.params.id;

    this.props.dispatch(editExpense(expense, id));
    this.props.history.push("/");
  };
  onRemove = () => {
    console.log("loading rm ....");
    console.log(this.props.match.params.id);

    this.props.dispatch(removeExpense({ id: this.props.match.params.id }));
    this.props.history.push("/");
  };
  render() {
    return (
      <div>
        <h2>EDITTING EXPENSE :</h2>
        <ExpenseForme expense={this.state.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onRemove}>remove</button>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
  };
};
const EditExpensePageConnected = connect(mapStateToProps)(EditExpensePage);

export default EditExpensePageConnected;
