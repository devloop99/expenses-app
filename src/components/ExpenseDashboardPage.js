import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseFilterInput from "./ExpenseFilterInput";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseFilterInput />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
