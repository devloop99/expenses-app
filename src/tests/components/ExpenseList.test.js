import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should expenselist render width expenses ", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});
test("should expenselist render without expenses ", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
