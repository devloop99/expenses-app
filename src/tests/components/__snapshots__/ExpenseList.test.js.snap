// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expenselist render width expenses  1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={2500}
    createdAt={0}
    description="rent"
    id={1}
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={200}
    createdAt={-1000}
    description="abonnement"
    id={2}
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={10000}
    createdAt={1000}
    description="car"
    id={3}
    key="3"
    note=""
  />
</div>
`;

exports[`should expenselist render without expenses  1`] = `
<div>
  <p>
    No expense yet ...
  </p>
</div>
`;
